src/main/native/cpp/support/hostname.cpp

```diff
@@ -0,0 +1,60 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) FIRST 2017. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "support/hostname.h"
+
+#ifdef _WIN32
+#include <Winsock2.h>
+#pragma comment(lib, "Ws2_32.lib")
+#else
+#include <unistd.h>
+#endif
+
+#include <string>
+#include "llvm/SmallVector.h"
+#include "llvm/StringRef.h"
+
+#ifdef _WIN32
+struct WSAHelper {
+  WSAHelper() {
+    WSAData wsaData;
+    WORD wVersionRequested = MAKEWORD(2, 2);
+    WSAStartup(wVersionRequested, &wsaData);
+  }
+  ~WSAHelper() { WSACleanup(); }
+};
+
+static WSAHelper& GetWSAHelper() {
+  static WSAHelper helper;
+  return helper;
+}
+#endif
+
+namespace wpi {
+std::string GetHostname() {
+#ifdef _WIN32
+  GetWSAHelper();
+#endif
+  char name[256];
+  if (::gethostname(name, sizeof(name)) != 0) return "";
+  name[255] = '\0';  // Per POSIX, may not be null terminated if too long
+  return name;
+}
+
+llvm::StringRef GetHostname(llvm::SmallVectorImpl<char>& name) {
+#ifdef _WIN32
+  GetWSAHelper();
+#endif
+  name.clear();
+  name.append(256, '\0');
```

Comment:
I think it'd be better to use a char array and then copy out of it.  That way if you're passed a SmallString<64> you won't hit the heap unless the actual hostname was that long.